File: 2024/day_02/puzzle.py
def read_data(filename):
    return open(f"{filename}").read().strip()

###############################
## Construct data for solver ##
###############################
def constructData(data):
    data = data.split("\n")

    map = []
    for i in range(len(data)):
        line = [int(x) for x in data[i].split(" ")] 
        map.append(line)
        
    return map


###########
## Utils ##
###########
def isSafe(line):
    safe = line[0] != line[1]
    if not(safe):
        return 1
    
    checkType = "ASC" if line[0] < line[1] else "DESC"
    
    for i in range(1, len(line)):
        if checkType == "ASC":
            safe = safe & ((line[i] - line[i-1]) > 0) & ((line[i] - line[i-1]) < 4)
        else:
            safe = safe & ((line[i-1] - line[i]) > 0) & ((line[i-1] - line[i]) < 4)

        if not(safe):
            return i

    return 0

###################
## Solver part 1 ##
###################
def part1(data):
    res = 0

    for l in range(len(data)):
        res += 1 if isSafe(data[l]) == 0 else 0

    return res


###################
## Solver part 2 ##
###################
def part2(data):
    res = 0

    for l in range(len(data)):
        if isSafe(data[l]) == 0:
            res += 1
        else:
            safe = False
            for i in range(len(data[l])):
                cp = data[l].copy()
                cp.pop(i)
                safe = safe or isSafe(cp)==0
            if safe:
                res += 1
    
    return res


##########
## MAIN ##
##########
data = read_data("puzzle_data.txt")
print('Result for puzzle 1 = ', part1(constructData(data)))
print('Result for puzzle 2 = ', part2(constructData(data)))

File: 2022/day_04/puzzle.js
var fs = require('fs');
require.extensions['.txt'] = function (module, filename) {
    module.exports = fs.readFileSync(filename, 'utf8');
};

const data_test = require('./data_test.txt');
const puzzle_data = require('./puzzle_data.txt');

function compute_data(data) {
    var res = []

    var split = data.split("\n");
    for (var i=0; i<split.length; i++) {
        let s = split[i]
        res.push(s.split(","))
    }
    return res;
}

function part1(data) {
    let res = 0
    for (let i=0; i<data.length; i++) {
        let first = data[i][0].split("-")
        let min1 = Number(first[0]), max1 = Number(first[1])
        let second = data[i][1].split("-")
        let min2 = Number(second[0]), max2 = Number(second[1])
        if (min1 <= min2 && max1 >= max2) {
            res++
        } else if (min1 >= min2 && max1 <= max2) {
            res++
        }
    }
    return res
}

function part2(data) {
    let res = 0
    for (let i=0; i<data.length; i++) {
        let first = data[i][0].split("-")
        let min1 = Number(first[0]), max1 = Number(first[1])
        let second = data[i][1].split("-")
        let min2 = Number(second[0]), max2 = Number(second[1])
        if (min2 <= min1 && min1 <= max2) {
            res++
        } else if (min2 <= max1 && max1 <= max2) {
            res++
        } else if (min1 <= min2 && min2 <= max1) {
            res++
        } else if (min1 <= max2 && max2 <= max1) {
            res++
        }
    }
    return res
}


let data = compute_data(puzzle_data)
console.log('Result for puzzle 1 = ', part1(data))
console.log('Result for puzzle 2 = ', part2(data))